snuba/utils/streams/processing.py

```diff
@@ -30,9 +30,25 @@ class ProcessingStrategy(ABC, Generic[TPayload]):
     """
 
     @abstractmethod
-    def process(self, message: Optional[Message[TPayload]]) -> None:
+    def poll(self) -> None:
         """
-        Process a message, or periodic heartbeat (``None`` value.)
+        Poll the processor.
+
+        This method is called on each consumer loop iteration. It can be used
+        to check on the status of asynchronous tasks or perform other
+        scheduled tasks.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def submit(self, message: Message[TPayload]) -> None:
+        """
+        Submit a message for processing.
+
+        Messages may be processed synchronously or asynchronously, depending
+        on the implementation of the processing strategy. Callers of this
+        method should not assume that this method returning successfully
+        implies that the message was successfully processed.
```

Comment:
Introducing asynchronous execution to these steps also introduces the problem of dealing with [backpressure](https://www.tedinski.com/2019/03/05/backpressure.html): steps that do not process messages serially but instead accumulate messages to process asynchronously will need to reject new messages being submitted when they have reached capacity, and upstream steps will need to be able to accommodate that. For example, the `ParallelTransform` will have a finite amount of shared memory that can be used to store message payloads: when this limit is reached, that step will need to start rejecting calls to `submit`.

This rejection can be done by making `submit` either a blocking call without a timeout, a blocking call with a timeout (and throwing if the timeout is reached and capacity hasn't become available), or a nonblocking call that throws an exception immediately if there is no available capacity — basically, the same behavior as `queue.put`. Blocking indefinitely  is problematic with Kafka due to the `max.poll.interval.ms` ([explanation here](https://stackoverflow.com/a/39759329)), so I think that either blocking with a timeout, or a nonblocking and throwing an exception are the best options available here.